url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/io/Buffer.java

```diff
@@ -0,0 +1,191 @@
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+//
+// Contact information:
+// Mailing address:
+// Marios Hadjieleftheriou
+// University of California, Riverside
+// Department of Computer Science
+// Surge Building, Room 310
+// Riverside, CA 92521
+//
+// Email:
+// marioh@cs.ucr.edu
+
+package spatialindex.io;
+
+import java.util.HashMap;
+import java.util.Iterator;
+
+/**
+ * use ram buffer to accelerate the operation
+ * 
+ */
+public abstract class Buffer implements IBuffer {
+	int m_capacity = 10;
+	boolean m_bWriteThrough = false;
+	IStorageManager m_storageManager = null;
+	HashMap<Integer, Entry> m_buffer = new HashMap<Integer, Entry>();
+	/**
+	 * record the hit number
+	 */
+	long m_hits = 0;
+
+	abstract void addEntry(int id, Entry entry);
+
+	abstract void removeEntry();
+
+	/**
+	 * be a buffer of sm with specific capacity
+	 * 
+	 * @param sm
+	 * @param capacity
+	 * @param bWriteThrough
+	 */
+	public Buffer(IStorageManager sm, int capacity, boolean bWriteThrough) {
+		m_storageManager = sm;
+		m_capacity = capacity;
+		m_bWriteThrough = bWriteThrough;
+	}
+
+	@Override
+	public byte[] loadByteArray(final int id) {
+		byte[] ret = null;
+		Entry e = m_buffer.get(new Integer(id));
+
+		if (e != null) {
+			m_hits++;
+
+			ret = new byte[e.m_data.length];
+			System.arraycopy(e.m_data, 0, ret, 0, e.m_data.length);
+		} else {
+			ret = m_storageManager.loadByteArray(id);
+			e = new Entry(ret);
+			addEntry(id, e);
+		}
+
+		return ret;
+	}
+
+	@Override
+	public int storeByteArray(final int id, final byte[] data) {
+		int ret = id;
+
+		if (id == NewPage) {
+			ret = m_storageManager.storeByteArray(id, data);
+			Entry e = new Entry(data);
+			addEntry(ret, e);
+		} else {
+			if (m_bWriteThrough) {
+				m_storageManager.storeByteArray(id, data);
+			}
+
+			Entry e = m_buffer.get(new Integer(id));
+			if (e != null) {
+				e.m_data = new byte[data.length];
+				System.arraycopy(data, 0, e.m_data, 0, data.length);
+
+				if (m_bWriteThrough == false) {
+					e.m_bDirty = true;
+					m_hits++;
+				} else {
+					e.m_bDirty = false;
+				}
+			} else {
+				e = new Entry(data);
+				if (m_bWriteThrough == false)
+					e.m_bDirty = true;
+				addEntry(id, e);
+			}
+		}
+
+		return ret;
+	}
+
+	@Override
+	public void deleteByteArray(final int id) {
+		Integer ID = new Integer(id);
+		Entry e = m_buffer.get(ID);
+		if (e != null) {
+			m_buffer.remove(ID);
+		}
+
+		m_storageManager.deleteByteArray(id);
+	}
+
+	@Override
+	/**
+	 * write dirty data to file
+	 */
+	public void flush() {
+		Iterator<java.util.Map.Entry<Integer, Entry>> it = m_buffer.entrySet()
+				.iterator();
+
+		while (it.hasNext()) {
+			java.util.Map.Entry<Integer, Entry> me = it.next();
+			Entry e = me.getValue();
+			int id = (me.getKey()).intValue();
+			if (e.m_bDirty)
+				m_storageManager.storeByteArray(id, e.m_data);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 139:145, 158:166.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.